Algorithms/Search/KmpSearcher.cs

```diff
@@ -0,0 +1,101 @@
+﻿using System.Collections.Generic;
+using Utilities.Exceptions;
+
+namespace Algorithms.Search
+{
+    /// <summary>
+    /// An implementation of Knuth–Morris–Pratt Algorithm.
+    /// Worst case time complexity: O(n).
+    /// </summary>
+    public class KmpSearcher
+    {
+        /// <summary>
+        /// Enumerates each zero-based index of all occurrences of the string in this instance.
+        /// throws ItemNotFoundException if the item couldn't be found.
+        /// </summary>
+        /// <param name="str">The string instance.</param>
+        /// <param name="pat">The pattern to seek.</param>
+        /// <returns>
+        /// The zero-based index positions of value if one or more <paramref name="pat"/> are found.
+        /// If <paramref name="pat"/> is empty, no indexes will be enumerated.
+        /// </returns>
+        /// <exception cref="ItemNotFoundException"><paramref name="str"/> or <paramref name="pat"/> is null.</exception>
+        public IEnumerable<int> FindIndexes(string str, string pat)
```

Comment:
Codacy found an issue: [Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6285773)